job-runner/chain/chain.go

```diff
@@ -14,34 +14,61 @@ import (
 )
 
 // chain represents a job chain and some meta information about it.
-type chain struct {
-	jcMux    *sync.RWMutex
+type Chain struct {
+	// For access to jobChain.Jobs map. Be careful not to make nested RLock()
+	// calls on jobsMux within the same goroutine.
+	jobsMux  *sync.RWMutex
 	jobChain *proto.JobChain
 
-	runningMux         *sync.RWMutex
-	running            map[string]proto.JobStatus // keyed on job id
-	n                  uint                       // N jobs ran
-	sequenceRetryCount map[string]uint            // Number of retries attempted so far
+	runningMux *sync.RWMutex
+	running    map[string]proto.JobStatus // keyed on job id
+	numJobsRun uint                       // Number of jobs run so far
+
+	triesMux          *sync.RWMutex   // for access to sequence/job tries maps
+	sequenceTries     map[string]uint // Number of sequence retries attempted so far
+	latestRunJobTries map[string]uint // job.Id -> number of times tried within the latest time it was run (i.e. within the latest sequence try)
+	totalJobTries     map[string]uint // job.Id -> total number of times tried
 }
 
-// NewChain takes a JobChain proto (from the RM) and turns it into a Chain that
-// the JR can use.
-func NewChain(jc *proto.JobChain) *chain {
-	sequenceRetryCount := make(map[string]uint)
+// NewChain takes a JobChain proto and maps of sequence + jobs trues, and turns them
+// into a Chain that the JR can use.
+func NewChain(jc *proto.JobChain, sequenceTries map[string]uint, totalJobTries map[string]uint, latestRunJobTries map[string]uint) *Chain {
+	// Make sure all jobs have valid State + Data fields, and count the number of
+	// completed + failed jobs (the number of jobs that have finished running).
+	numJobsRun := uint(0)
 	for jobName, job := range jc.Jobs {
-		job.State = proto.STATE_PENDING
-		job.Data = map[string]interface{}{}
+		switch job.State {
+		case proto.STATE_PENDING:
+			// Pending is a valid job state - do nothing.
+		case proto.STATE_COMPLETE:
+			// Valid state, job is done running.
+			numJobsRun += 1
+		case proto.STATE_FAIL:
+			// Valid state, job is done running.
+			numJobsRun += 1
+		default:
+			// Job isn't pending, failed, or complete. For a new / suspended chain,
+			// these are the only valid states (no jobs can be running before the
+			// chain is started or resumed). Set jobs with other states to pending.
```

Comment:
So jobs with STATE_TIMEOUT and STATE_STOPPED will both be set to STATE_PENDING, right? Just want to make sure that's intended

Reply:
Ah, good catch, actually Stopped jobs need to stay STATE_STOPPED so we get the number of job tries right later on. Changed so Stopped jobs stay Stopped and all other states (Running, Timeout) become Failed instead of Pending.